packages/flutter/test/widgets/grid_view_test.dart

```diff
@@ -899,4 +899,134 @@ void main() {
+      child: StatefulBuilder(
+        builder: (BuildContext context, StateSetter setState) {
+          stateSetter = setState;
+          return GridView.builder(
+            gridDelegate: SliverGridDelegateWithFixedCrossAxisCount(
+              crossAxisCount: 3,
+              mainAxisSpacing: 8,
+              crossAxisSpacing: 8,
+              mainAxisExtent: mainAxisExtent,
+            ),
+            itemBuilder: (BuildContext context, int index) {
+              return Container(
+                height: 50,
+                alignment: Alignment.center,
+                child: Text('$index'),
+              );
+            },
+          );
+        },
+      ),
+    ));
+    expect(tester.takeException(), isNull);
+
+    stateSetter(() => mainAxisExtent = -100);
+    await tester.pump();
+
+    expect(tester.takeException(), isNull);
+  });
+
+  testWidgets(
+      'SliverGridDelegateWithMaxCrossAxisExtent mainAxisExtent corrected',
+      (WidgetTester tester) async {
+    // Regression test for https://github.com/flutter/flutter/issues/138871
+    const SliverGridDelegateWithMaxCrossAxisExtent delegate =
+        SliverGridDelegateWithMaxCrossAxisExtent(
+      maxCrossAxisExtent: 100,
+      mainAxisSpacing: 8,
+      crossAxisSpacing: 8,
+      mainAxisExtent: -100,
+    );
+
+    final SliverGridRegularTileLayout sliverGridRegularTileLayout =
+        delegate.getLayout(
+      const SliverConstraints(
+        axisDirection: AxisDirection.down,
+        growthDirection: GrowthDirection.forward,
+        userScrollDirection: ScrollDirection.forward,
+        scrollOffset: 100.0,
+        precedingScrollExtent: 0.0,
+        overlap: 0.0,
+        remainingPaintExtent: 0.0,
+        crossAxisExtent: 500,
+        crossAxisDirection: AxisDirection.right,
+        viewportMainAxisExtent: 100.0,
+        remainingCacheExtent: 0.0,
+        cacheOrigin: 0.0,
+      ),
+    ) as SliverGridRegularTileLayout;
+
+    expect(sliverGridRegularTileLayout.childMainAxisExtent, 0);
+
+    double mainAxisExtent = 100;
+    late StateSetter stateSetter;
+    await tester.pumpWidget(Directionality(
+      textDirection: TextDirection.ltr,
+      child: StatefulBuilder(
+        builder: (BuildContext context, StateSetter setState) {
+          stateSetter = setState;
+          return GridView.builder(
+            gridDelegate: SliverGridDelegateWithMaxCrossAxisExtent(
+              maxCrossAxisExtent: 100,
+              mainAxisSpacing: 8,
+              crossAxisSpacing: 8,
+              mainAxisExtent: mainAxisExtent,
+            ),
+            itemBuilder: (BuildContext context, int index) {
+              return Container(
+                height: 50,
+                alignment: Alignment.center,
+                child: Text('$index'),
+              );
+            },
+          );
+        },
+      ),
+    ));
+    expect(tester.takeException(), isNull);
+
+    stateSetter(() => mainAxisExtent = -100);
```

Comment:
I don't think this updates the widget. It just changes the variable value. I would expect a main axis extent of this value would throw an exception since it is not valid.